File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.py
class Solution:
    def smallestPalindrome(self, s: str) -> str:
        freq = [0] * 26

        # Count frequency
        for ch in s:
            freq[ord(ch) - ord('a')] += 1

        left = []
        middle = ""

        # Build left half
        for i in range(26):
            left.extend(chr(ord('a') + i) * (freq[i] // 2))

            if freq[i] % 2 == 1:
                middle = chr(ord('a') + i)

        left = "".join(left)
        right = left[::-1]

        return left + middle + right

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.cpp
// class Solution {
// public:
//     string smallestPalindrome(string s) {

//         vector<int> freq(26, 0);

//         for (char ch : s)
//             freq[ch - 'a']++;

//         string left = "";
//         string mid = "";

//         for (int i = 0; i < 26; i++) {

//             left += string(freq[i] / 2, char('a' + i));

//             if (freq[i] % 2 == 1)
//                 mid = char('a' + i);
//         }

//         string right = left;
//         reverse(right.begin(), right.end());

//         return left + mid + right;
//     }
// };
class Solution {
public:
    string smallestPalindrome(string s) {
        int n = s.size();
        int counts[26] = {0};
        
        // 1. O(N) 
        for(int i = 0; i < n / 2; ++i) {
            counts[s[i] - 'a']++;
        }
        
        // 2. O(N)
        int idx = 0;
        for(int i = 0; i < 26; ++i) {
            while(counts[i]-- > 0) {
                s[idx++] = (char)(i + 'a');
            }
        }
        
        // 3. O(N)
        for(int i = 0; i < n / 2; ++i) {
            s[n - 1 - i] = s[i];
        }
        
        return s;
    }
};